src/rust/ensogl/lib/text/src/buffer/data/text.rs

```diff
@@ -0,0 +1,762 @@
+    pub fn last_line_byte_offset(&self) -> Bytes {
+        self.byte_offset_from_line_index_unchecked(self.last_line_index())
+    }
+
+    /// The start column of the last line.
+    pub fn last_line_start_column(&self) -> Column {
+        0.column()
+    }
+
+    /// The start location of the last line.
+    pub fn last_line_start_location(&self) -> Location {
+        let line   = self.last_line_index();
+        let column = self.last_line_start_column();
+        Location(line,column)
+    }
+
+    /// The last column number of the last line.
+    pub fn last_line_end_column(&self) -> Column {
+        self.column_from_byte_offset(self.byte_size()).unwrap()
+    }
+
+    /// The byte offset of the end of the last line. Equal to the byte size of the whole text.
+    pub fn last_line_end_byte_offset(&self) -> Bytes {
+        self.byte_size()
+    }
+
+    /// The location of the last character in the text.
+    pub fn last_line_end_location(&self) -> Location {
+        let line   = self.last_line_index();
+        let column = self.last_line_end_column();
+        Location(line,column)
+    }
+}
+
+
+// === Validation ===
+
+impl Text {
+    /// Check whether the provided line index is valid in this text.
+    pub fn validate_line_index(&self, line:Line) -> Result<Line,BoundsError> {
+        use BoundsError::*;
+        if      line < 0.line()               {Err(TooSmall)}
+        else if line > self.last_line_index() {Err(TooBig)}
+        else                                  {Ok(line)}
+    }
+
+    /// Check whether the provided byte offset is valid in this text.
+    pub fn validate_byte_offset(&self, offset:Bytes) -> Result<Bytes,BoundsError> {
+        use BoundsError::*;
+        if      offset < 0.bytes()        {Err(TooSmall)}
+        else if offset > self.byte_size() {Err(TooBig)}
+        else                              {Ok(offset)}
+    }
+}
+
+
+
+// ===================
+// === Conversions ===
+// ===================
+
+// === Into Byte Offset ===
+
+impl Text {
+    /// Return the offset after the last character of a given line if the line exists.
+    pub fn end_byte_offset_from_line_index(&self, line:Line) -> Result<Bytes,BoundsError> {
+        self.validate_line_index(line)?;
+        let next_line      = line + 1.line();
+        let next_line_off  = self.byte_offset_from_line_index(next_line).ok();
+        let next_line_prev = next_line_off.and_then(|t| self.prev_grapheme_offset(t));
+        Ok(next_line_prev.unwrap_or_else(|| self.byte_size()))
+    }
+
+    /// Return the offset after the last character of a given line if the line exists. Snapped to
+    /// the closest valid value.
+    pub fn end_byte_offset_from_line_index_snapped(&self, line:Line) -> Bytes {
+        self.snap_bytes_bounds_result(self.end_byte_offset_from_line_index(line))
+    }
+
+    /// The line byte offset. Panics in case the line index was invalid.
+    pub fn byte_offset_from_line_index_unchecked(&self, line:Line) -> Bytes {
```

Comment:
Few reasons why I dont think this is good idea:

1. When creating names we should put in the name head the most important part, especially when there is so many functions associated with this struct. What we want to compute here is the `byte_offset`, so it should be the first name of the function. Maybe in the future we will make it more generic and we would allow to just use `self.byte_offset(...)` with any paraeter? The `_from_line_index` part is just repeating the types, so its not really important in this name.
2. If we really want to change the names, we need to have bigger discussion here. For example, right now we have such names: `byte_offset_from_line_index`, and `line_index_from_byte_offset`. Would you change them to `line_byte_offset` and `byte_offset_line`? For me, the later forms are really hard to understand. There is currently over 40 (!) such conversion functions. I was trying to assign nicer names to them on the beginning and I refactored everything later to a consistent naming schema.

From the above points, point 1 is more important for me and I believe we should always design names this way - if a name contains unnecessary part (like repetition of the name of argument type), this should be put at the end of the name, or, if we have more time to do it, the functions should be defined as polymorphic functions without using this suffix at all.

Reply:
For me the main principle is "I understand the function name as it is written in english". `line_byte_offset` is actually bad, because you cannot put `'s` into name, but what about `byte_offset_of_line`?